### PR TITLE
Automatically guard when SymInt is converted to int

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95479

During enablement, we disabled int() conversions because they were
any easy way to footgun guards.  We have enough of dynamic shapes
working now that this is now causing spurious errors; e.g., if you feed
a symbolic int to x.size(symint).  We now allow for implicit conversions
of SymInt to int here, posting a guard.  We expect guard provenance
to help people debug overspecialization.

Fixes https://github.com/pytorch/pytorch/issues/95328

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: test/dynamo/test_repros.py [hunk 1]
```diff
@@ -2335,6 +2335,19 @@ def f(x):
...
 
+    @torch._dynamo.config.patch("dynamic_shapes", True)
+    def test_dynamic_shapes_implicit_guard(self):
+        def f(x):
+            y = x * x.size(x.shape[0])
+            torch.sum(y, [y.shape[0]])
+            return y
+
+        cnt = torch._dynamo.testing.CompileCounter()
+        opt_fn = torch._dynamo.optimize(cnt, nopython=True)(f)
```
Comment (wconstab) [inline test/dynamo/test_repros.py:2346]: is your test only caring about nopython not raising?

i thought i'd see a second call to opt_fn and a recompile to prove the guard worked

File: torch/csrc/utils/python_arg_parser.h [hunk 2]
```diff
@@ -944,6 +947,9 @@ inline c10::optional<double> PythonArgs::toDoubleOptional(int i) {
 inline double PythonArgs::toDouble(int i) {
   if (!args[i])
     return signature.params[i].default_double;
+  if (torch::is_symfloat(py::handle(args[i]))) {
```
Comment (wconstab) [inline torch/csrc/utils/python_arg_parser.h:950]: hmm also float?

File: test/dynamo/test_repros.py (hunk 1 above)
Comment (ezyang) [reply test/dynamo/test_repros.py:2346]: I mostly only wanted to show that the test got to the end without erroring or graph breaking. I wasn't intending to exercise the guard machinery, although I suppose I could add a second call if you really want

Comment (wconstab) [reply test/dynamo/test_repros.py:2346]: no need, just wasn't sure what you were testing

**Review by wconstab — APPROVED** [review]
This seems good, and we can always change it...

File: torch/csrc/utils/python_arg_parser.h (hunk 2 above)
Comment (ezyang) [reply torch/csrc/utils/python_arg_parser.h:950]: There isn't a toFloat :P

Comment (ngimel) [reply torch/csrc/utils/python_arg_parser.h:951]: This path is not tested?

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -402,11 +402,8 @@ def sym_and(self, other):  # noqa: F811
     def is_non_overlapping_and_dense(self, sizes, strides):
         return self.is_non_overlapping_and_dense_indicator(sizes, strides).eq(to_node(self, 1))  # type: ignore[attr-defined]
 
-    # Today we error on calling int on a symbolic shape, as this is a very accessible footgun.
     def int_(self):
```
Comment (Skylion007) [inline torch/fx/experimental/symbolic_shapes.py:405]: Ooh, would implementing `__int__` someday be in the cards? Or is there a really good reason we don't ever want that.

**Review by voznesenskym — APPROVED** [review]

File: torch/fx/experimental/symbolic_shapes.py (hunk 3 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:405]: This PR also turns on `__int__` on SymInt. SymNode doesn't have any magic methods on it though.

Comment (Skylion007) [reply torch/fx/experimental/symbolic_shapes.py:405]: Would it make sense for SymInt to subclass `numbers.Integral` then?

**Review by Chillee — COMMENTED** [review]
makes sense to me.

Will we just always guard instead of erroring now?

File: torch/fx/experimental/symbolic_shapes.py (hunk 3 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:405]: Yeah, we can do that

**Comment by ezyang** [comment]
Yup

File: torch/csrc/utils/python_arg_parser.h (hunk 2 above)
Comment (ezyang) [reply torch/csrc/utils/python_arg_parser.h:951]: Added!

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**